device-pkgs/initrdflash-script.nix

```diff
@@ -0,0 +1,141 @@
+              */
+            usb-role-switch;
+            connector {
+              compatible = "usb-b-connector", "gpio-usb-b-connector";
+              label = "usb-recovery";
+              cable-connected-on-boot = <2>;
+            };
+          };
+        };
+
+        fragment@1 {
+          target-path = "/${xudcPath}";
+
+          board_config {
+            sw-modules = "kernel", "uefi";
+          };
+
+          __overlay__ {
+            status = "okay";
+          };
+        };
+      };
+    '';
+  forceXusbPeripheralDtbo = deviceTree.compileDTS {
+    # We force USB port to peripheral mode, in case the carrier board doesn't support OTG
+    # Fortunately, the USB controller is all in the SoC, so we can override it directly here
+    name = "force-xusb-peripheral.dtbo";
+    dtsFile = forceXusbPeripheralDts;
+  };
+
+  rcmScript = mkRcmBootScript {
+    kernelPath = "${config.system.build.kernel}/${config.system.boot.loader.kernelFile}";
+    initrdPath = "${flashInitrd}/initrd";
+    kernelCmdline = lib.concatStringsSep " " [
+      "console=ttyTCU0,115200"
+      "sdhci_tegra.en_boot_part_access=1"
+    ];
+    # During the initrd flash script, we upload two edk2 builds to the
+    # board, one that is only used temporarily to boot into our
+    # kernel/initrd to perform the flashing, and another one that is
+    # persisted to the firmware storage medium for future booting. We
+    # don't want to influence the boot order of the temporary edk2 since
+    # this may cause that edk2 to boot from something other than our
+    # intended flashing kernel/initrd combo (e.g. disk or network). Since
+    # the edk2 build that we actually persist to the board is embedded in
+    # the initrd used for flashing, we have the desired boot order (as
+    # configured in nix) in there and is not affected dynamically during
+    # the flashing procedure. NVIDIA ensures that when the device is
+    # using RCM boot, only the boot mode named "boot.img" is used (see https://gist.github.com/jmbaur/1ca79436e69eadc0a38ec0b43b16cb2f#file-flash-sh-L1675).
+    additionalDtbOverlays = (lib.filter (path: (path.name or "") != "DefaultBootOrder.dtbo") cfg.flashScriptOverrides.additionalDtbOverlays) ++ [ forceXusbPeripheralDtbo ];
+  };
+in
+''
+  ${rcmScript}
+
+  echo
+  echo "Jetson device should now be flashing and will reboot when complete."
+
+  # TODO(eberman): maybe stop assuming /dev/serial/by-id is present
+  # It'll be populated with udev, so most Linux systems will work fine
+  echo -n "Waiting for Jetson device to appear via USB, this may take up to 4 minutes"
+
+  counter=0
+  until [[ -e /dev/serial/by-id/${serialPortId} || $counter -gt ${builtins.toString (4 * 60 * 2)} ]] ; do
+    echo -n "."
+    sleep 0.5
+    ((++counter))
+  done
+  echo
+
+  if [ ! -e /dev/serial/by-id/${serialPortId} ] ; then
+    echo "Jetson device not detected, please monitor the serial port or wait 5-15 minutes"
+    exit 2
+  fi
+
+  logs=$(mktemp)
```

Comment:
I'd prefer if we didn't have to have a temporary log file here, and either use a pipe or "expect" instead.